File: charging-pile/src/frame.cpp
#include "frame.h"

#include <ctime>
#include <stdexcept>

namespace frame {

namespace {
constexpr uint8_t START = 0x68;
constexpr std::size_t HEADER_LEN = 14;  // 起始..主机编号末
constexpr std::size_t FCS_LEN = 2;
constexpr std::size_t MAX_PAYLOAD = 8 * 1024;

int hex_val(char c) {
    if (c >= '0' && c <= '9') return c - '0';
    if (c >= 'a' && c <= 'f') return c - 'a' + 10;
    if (c >= 'A' && c <= 'F') return c - 'A' + 10;
    throw std::runtime_error("非法 hex 字符");
}
}  // namespace

std::vector<unsigned char> hex_to_bytes(const std::string& hex) {
    if (hex.size() % 2 != 0) {
        throw std::runtime_error("hex 长度非偶");
    }
    std::vector<unsigned char> out;
    out.reserve(hex.size() / 2);
    for (std::size_t i = 0; i < hex.size(); i += 2) {
        out.push_back(static_cast<unsigned char>((hex_val(hex[i]) << 4) | hex_val(hex[i + 1])));
    }
    return out;
}

std::string bytes_to_hex(const std::vector<unsigned char>& data) {
    static const char* d = "0123456789abcdef";
    std::string out;
    out.reserve(data.size() * 2);
    for (unsigned char b : data) {
        out.push_back(d[b >> 4]);
        out.push_back(d[b & 0x0f]);
    }
    return out;
}

// CRC-16/CCITT-FALSE：poly=0x1021, init=0xFFFF, 不反转, xorout=0
uint16_t crc16_ccitt(const std::vector<unsigned char>& data) {
    uint16_t crc = 0xFFFF;
    for (unsigned char b : data) {
        crc ^= static_cast<uint16_t>(b) << 8;
        for (int i = 0; i < 8; i++) {
            if (crc & 0x8000) {
                crc = static_cast<uint16_t>((crc << 1) ^ 0x1021);
            } else {
                crc = static_cast<uint16_t>(crc << 1);
            }
        }
    }
    return crc;
}

// 主机编号 → 7 字节 BCD：取 id 中的数字，右对齐成 14 位十进制再压 BCD
std::vector<unsigned char> bcd7_from_id(const std::string& id) {
    std::string digits;
    for (char c : id) {
        if (c >= '0' && c <= '9') digits.push_back(c);
    }
    if (digits.size() > 14) {
        digits = digits.substr(digits.size() - 14);
    }
    while (digits.size() < 14) {
        digits.insert(digits.begin(), '0');
    }
    std::vector<unsigned char> out(7, 0);
    for (std::size_t i = 0; i < 7; i++) {
        out[i] = static_cast<unsigned char>(((digits[2 * i] - '0') << 4) | (digits[2 * i + 1] - '0'));
    }
    return out;
}

// CP56Time2a：毫秒(2) 分(1) 时(1) 日(1) 月(1) 年(1)
std::vector<unsigned char> cp56_now() {
    std::time_t t = std::time(nullptr);
    std::tm tm{};
    localtime_r(&t, &tm);
    unsigned int ms = static_cast<unsigned int>(tm.tm_sec) * 1000;
    std::vector<unsigned char> out(7, 0);
    out[0] = static_cast<unsigned char>(ms & 0xff);
    out[1] = static_cast<unsigned char>((ms >> 8) & 0xff);
    out[2] = static_cast<unsigned char>(tm.tm_min & 0x3f);
    out[3] = static_cast<unsigned char>(tm.tm_hour & 0x1f);
    out[4] = static_cast<unsigned char>(tm.tm_mday & 0x1f);
    out[5] = static_cast<unsigned char>((tm.tm_mon + 1) & 0x0f);
    out[6] = static_cast<unsigned char>(tm.tm_year % 100);
    return out;
}

std::vector<unsigned char> encode(uint8_t type, uint16_t seq,
                                  const std::vector<unsigned char>& host_no,
                                  const std::vector<unsigned char>& payload) {
    if (host_no.size() != 7) {
        throw std::runtime_error("主机编号必须 7 字节");
    }
    if (payload.size() > MAX_PAYLOAD) {
        throw std::runtime_error("载荷超长");
    }
    std::vector<unsigned char> f;
    f.reserve(HEADER_LEN + payload.size() + FCS_LEN);
    f.push_back(START);
    f.push_back(static_cast<unsigned char>((payload.size() >> 8) & 0xff));
    f.push_back(static_cast<unsigned char>(payload.size() & 0xff));
    f.push_back(static_cast<unsigned char>((seq >> 8) & 0xff));
    f.push_back(static_cast<unsigned char>(seq & 0xff));
    f.push_back(0x00);  // 加密标志：协商阶段明文
    f.push_back(type);
    f.insert(f.end(), host_no.begin(), host_no.end());
    f.insert(f.end(), payload.begin(), payload.end());
    uint16_t crc = crc16_ccitt(f);
    f.push_back(static_cast<unsigned char>((crc >> 8) & 0xff));
    f.push_back(static_cast<unsigned char>(crc & 0xff));
    return f;
}

Frame read(NetClient& net) {
    auto head = net.read_exact(HEADER_LEN);
    if (head[0] != START) {
        throw std::runtime_error("帧起始符非法");
    }
    std::size_t plen = (static_cast<std::size_t>(head[1]) << 8) | head[2];
    if (plen > MAX_PAYLOAD) {
        throw std::runtime_error("帧载荷超长");
    }
    auto payload = net.read_exact(plen);
    auto fcs = net.read_exact(FCS_LEN);

    std::vector<unsigned char> covered = head;
    covered.insert(covered.end(), payload.begin(), payload.end());
    uint16_t got = static_cast<uint16_t>((fcs[0] << 8) | fcs[1]);
    if (got != crc16_ccitt(covered)) {
        throw std::runtime_error("帧校验(CRC16)失败");
    }

    Frame fr;
    fr.seq = static_cast<uint16_t>((head[3] << 8) | head[4]);
    fr.enc_flag = head[5];
    fr.type = head[6];
    fr.host_no.assign(head.begin() + 7, head.begin() + 14);
    fr.payload = std::move(payload);
    return fr;
}

uint8_t Reader::u8() {
    if (pos_ + 1 > buf_.size()) {
        throw std::runtime_error("载荷读取越界(u8)");
    }
    return buf_[pos_++];
}

std::vector<unsigned char> Reader::take(std::size_t n) {
    if (pos_ + n > buf_.size()) {
        throw std::runtime_error("载荷读取越界");
    }
    std::vector<unsigned char> out(buf_.begin() + pos_, buf_.begin() + pos_ + n);
    pos_ += n;
    return out;
}

std::string Reader::take_hex(std::size_t n) {
    return bytes_to_hex(take(n));
}

}  // namespace frame

File: cloud-service/src/main/java/com/clpkc/cloud/socket/Frame.java
package com.clpkc.cloud.socket;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.time.LocalDateTime;
import java.util.Arrays;

/**
 * 桩 ↔ 云 二进制帧编解码（与 charging-pile/src/frame.{h,cpp} 字节级对齐）。
 * 帧结构与假设见 docs/WIRE_PROTOCOL.md：
 *   68 | 数据长度(2) | 序列号(2) | 加密标志(1) | 类型(1) | 主机编号(7 BCD) | 载荷 | CRC16(2)
 * 多字节大端；数据长度=载荷字节数；CRC-16/CCITT-FALSE 覆盖 [起始..载荷末]。
 */
final class Frame {

    // 报文类型
    static final int TYPE_P1_UP = 0x39;    // 一阶段 桩→云
    static final int TYPE_P1_DOWN = 0x3A;  // 一阶段 云→桩
    static final int TYPE_P2_UP = 0x3B;    // 二阶段 桩→云
    static final int TYPE_P2_DOWN = 0x3C;  // 二阶段 云→桩

    // 步骤号 STEP = 高4位轮次 | 低4位类别(1=请求,2=应答/回执)
    static final int STEP_1_REQ = 0x11;
    static final int STEP_1_RESP = 0x12;
    static final int STEP_2_REQ = 0x21;
    static final int STEP_2_RESP = 0x22;
    static final int STEP_3_REQ = 0x31;
    static final int STEP_3_RESP = 0x32;

    static final int RESULT_OK = 0x00;
    static final int RESULT_FAIL = 0x01;
    static final int RECEIVED_YES = 0x01;

    private static final int START = 0x68;
    private static final int HEADER_LEN = 14;
    private static final int FCS_LEN = 2;
    private static final int MAX_PAYLOAD = 8 * 1024;

    final int type;
    final int seq;
    final int encFlag;
    final byte[] hostNo;   // 7 字节 BCD
    final byte[] payload;  // 首字节为 STEP

    private Frame(int type, int seq, int encFlag, byte[] hostNo, byte[] payload) {
        this.type = type;
        this.seq = seq;
        this.encFlag = encFlag;
        this.hostNo = hostNo;
        this.payload = payload;
    }

    // ---- CRC-16/CCITT-FALSE：poly=0x1021, init=0xFFFF, 不反转, xorout=0 ----
    static int crc16(byte[] data, int len) {
        int crc = 0xFFFF;
        for (int i = 0; i < len; i++) {
            crc ^= (data[i] & 0xff) << 8;
            for (int b = 0; b < 8; b++) {
                if ((crc & 0x8000) != 0) {
                    crc = (crc << 1) ^ 0x1021;
                } else {
                    crc <<= 1;
                }
                crc &= 0xFFFF;
            }
        }
        return crc & 0xFFFF;
    }

    // 主机编号 → 7 字节 BCD：取 id 中的数字，右对齐成 14 位十进制再压 BCD
    static byte[] bcd7FromId(String id) {
        StringBuilder digits = new StringBuilder();
        for (int i = 0; i < id.length(); i++) {
            char c = id.charAt(i);
            if (c >= '0' && c <= '9') {
                digits.append(c);
            }
        }
        String s = digits.toString();
        if (s.length() > 14) {
            s = s.substring(s.length() - 14);
        }
        while (s.length() < 14) {
            s = "0" + s;
        }
        byte[] out = new byte[7];
        for (int i = 0; i < 7; i++) {
            out[i] = (byte) (((s.charAt(2 * i) - '0') << 4) | (s.charAt(2 * i + 1) - '0'));
        }
        return out;
    }

    // CP56Time2a：毫秒(2) 分(1) 时(1) 日(1) 月(1) 年(1)
    static byte[] cp56Now() {
        LocalDateTime now = LocalDateTime.now();
        int ms = now.getSecond() * 1000;
        byte[] out = new byte[7];
        out[0] = (byte) (ms & 0xff);
        out[1] = (byte) ((ms >> 8) & 0xff);
        out[2] = (byte) (now.getMinute() & 0x3f);
        out[3] = (byte) (now.getHour() & 0x1f);
        out[4] = (byte) (now.getDayOfMonth() & 0x1f);
        out[5] = (byte) (now.getMonthValue() & 0x0f);
        out[6] = (byte) (now.getYear() % 100);
        return out;
    }

    /** 组帧：type + seq + hostNo + payload → 完整帧字节。 */
    static byte[] encode(int type, int seq, byte[] hostNo, byte[] payload) {
        if (hostNo.length != 7) {
            throw new IllegalArgumentException("主机编号必须 7 字节");
        }
        if (payload.length > MAX_PAYLOAD) {
            throw new IllegalArgumentException("载荷超长");
        }
        byte[] f = new byte[HEADER_LEN + payload.length + FCS_LEN];
        int i = 0;
        f[i++] = (byte) START;
        f[i++] = (byte) ((payload.length >> 8) & 0xff);
        f[i++] = (byte) (payload.length & 0xff);
        f[i++] = (byte) ((seq >> 8) & 0xff);
        f[i++] = (byte) (seq & 0xff);
        f[i++] = 0x00;  // 加密标志：协商阶段明文
        f[i++] = (byte) type;
        System.arraycopy(hostNo, 0, f, i, 7);
        i += 7;
        System.arraycopy(payload, 0, f, i, payload.length);
        i += payload.length;
        int crc = crc16(f, i);
        f[i++] = (byte) ((crc >> 8) & 0xff);
        f[i] = (byte) (crc & 0xff);
        return f;
    }

    /** 从流读一帧（校验起始符与 CRC；失败抛 IOException）。 */
    static Frame read(InputStream in) throws IOException {
        byte[] head = readExact(in, HEADER_LEN);
        if ((head[0] & 0xff) != START) {
            throw new IOException("帧起始符非法");
        }
        int plen = ((head[1] & 0xff) << 8) | (head[2] & 0xff);
        if (plen > MAX_PAYLOAD) {
            throw new IOException("帧载荷超长");
        }
        byte[] payload = readExact(in, plen);
        byte[] fcs = readExact(in, FCS_LEN);

        byte[] covered = new byte[HEADER_LEN + plen];
        System.arraycopy(head, 0, covered, 0, HEADER_LEN);
        System.arraycopy(payload, 0, covered, HEADER_LEN, plen);
        int got = ((fcs[0] & 0xff) << 8) | (fcs[1] & 0xff);
        if (got != crc16(covered, covered.length)) {
            throw new IOException("帧校验(CRC16)失败");
        }
        int seq = ((head[3] & 0xff) << 8) | (head[4] & 0xff);
        int enc = head[5] & 0xff;
        int type = head[6] & 0xff;
        byte[] hostNo = Arrays.copyOfRange(head, 7, 14);
        return new Frame(type, seq, enc, hostNo, payload);
    }

    static void write(OutputStream out, byte[] frame) throws IOException {
        out.write(frame);
        out.flush();
    }

    private static byte[] readExact(InputStream in, int len) throws IOException {
        byte[] out = new byte[len];
        int off = 0;
        while (off < len) {
            int n = in.read(out, off, len - off);
            if (n < 0) {
                throw new IOException("连接被对端关闭");
            }
            off += n;
        }
        return out;
    }

    /** 载荷顺序读取器（定长字段直拼）。 */
    static final class Reader {
        private final byte[] buf;
        private int pos;

        Reader(byte[] buf) {
            this.buf = buf;
        }

        int u8() {
            if (pos + 1 > buf.length) {
                throw new IllegalStateException("载荷读取越界(u8)");
            }
            return buf[pos++] & 0xff;
        }

        byte[] take(int n) {
            if (pos + n > buf.length) {
                throw new IllegalStateException("载荷读取越界");
            }
            byte[] out = Arrays.copyOfRange(buf, pos, pos + n);
            pos += n;
            return out;
        }
    }

    /** 载荷顺序写入器。 */
    static final class Writer {
        private final ByteArrayOutputStream buf = new ByteArrayOutputStream(256);

        Writer step(int step) {
            buf.write(step & 0xff);
            return this;
        }

        Writer u8(int v) {
            buf.write(v & 0xff);
            return this;
        }

        Writer bytes(byte[] b) {
            buf.write(b, 0, b.length);
            return this;
        }

        byte[] toArray() {
            return buf.toByteArray();
        }
    }
}
